AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/GetKeys.dfy

```diff
@@ -203,46 +223,21 @@ module GetKeys {
         EncryptionContext := hv2EC,
         CiphertextBlob := branchKeyItemFromStorage.CiphertextBlob
       );
+      print "\nHv2 Branch Key: ";
+      print hv2BranchKey;
       var branchKey: KMS.DecryptResponse :- KMSKeystoreOperations.DecryptKeyForHV2(
         hv2BranchKey,
         kmsConfiguration,
         grantTokens,
         kmsClient
       );
-      var mdDigestMap := HierarchicalVersionUtils.GetMdDigestFromEC(branchKeyItemFromStorage.EncryptionContext);
-      var utf8MDDigest :- HierarchicalVersionUtils.UnstringifyEncryptionContext(mdDigestMap);
-      var crypto := HierarchicalVersionUtils.ProvideCryptoClient();
-      if (crypto.Failure?) {
-        var e := Types.KeyStoreException(
-          message :=
-            "Local Cryptography error: " + AtomicPrimitives.ErrorUtils.MessageOrUnknown(crypto.error));
-        return Failure(e);
-      }
-      var mdDigestSha := CanonicalEncryptionContext.EncryptionContextDigest(crypto.value, utf8MDDigest);
-      if (mdDigestSha.Failure?) {
-        var e := Types.KeyStoreException(
-          message :=
-            "Failed to create SHA of MD Digest.");
-        return Failure(e);
-      }
-      var plaintextBranchKeyWithMdDigest := branchKey.Plaintext.value;
-      :- Need(
-        |plaintextBranchKeyWithMdDigest| == Structure.AES_256_LENGTH + Structure.MD_DIGEST_LENGTH,
-        Types.KeyStoreException(
-          message := ErrorMessages.INVALID_HIERARCHY_VERSION
-        )
-      );
-      var plaintextBranchKey := plaintextBranchKeyWithMdDigest[0..Structure.AES_256_LENGTH];
-      var decryptedMdDigest := plaintextBranchKeyWithMdDigest[Structure.AES_256_LENGTH..];
-      if (decryptedMdDigest != mdDigestSha.value) {
-        var e := Types.KeyStoreException(
-          message :=
-            ErrorMessages.MD_DIGEST_SHA_NOT_MATCHED);
-        return Failure(e);
+      var validateResult := HierarchicalVersionUtils.ValidateMdDigest(branchKey.Plaintext.value, branchKeyItemFromStorage);
+      if (validateResult.Failure?) {
+        return Failure(validateResult.error);
       }
       var branchKeyMaterials :- Structure.ToBranchKeyMaterials(
         branchKeyItemFromStorage,
-        plaintextBranchKey
+        branchKey.Plaintext.value[Structure.MD_DIGEST_LENGTH..]
```

Comment:
Blocking: We need to validate the PDK is the right length before returning it.
Break the PDK out before calling `Structure.ToBranchKeyMaterials`. 
I get that other logic in the current code path 
ensures that the PDK is the right length today;
but it is fragile to assume that will always be the case.

Reply:
This is done inside ValidateMdDigest because we also need valid length to validate MD Digest

Reply:
> Break the PDK out before calling Structure.ToBranchKeyMaterials.

Added a TODO

Reply:
https://github.com/aws/aws-cryptographic-material-providers-library/pull/1354/files#r2011022778